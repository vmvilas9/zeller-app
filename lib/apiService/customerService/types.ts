export type ZellerCustomer =  {
    id: string;
    name: string;
    email: string;
    role: string;
}

export type ZellerCustomerConnection = {
    items: ZellerCustomer[];
    nextToken: string;
}

export type TableStringFilterInput = {
    ne?: string;
    eq?: string;
    le?: string;
    lt?: string;
    ge?: string;
    gt?: string;
    contains?: string;
    notContains?: string;
    between?: [string, string];
    beginsWith?: string;
};

export type TableZellerCustomerFilterInput = {
    id?: TableStringFilterInput;
    name?: TableStringFilterInput;
    email?: TableStringFilterInput;
    role?: TableStringFilterInput;
};
