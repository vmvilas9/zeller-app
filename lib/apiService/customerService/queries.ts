import { gql } from "@apollo/client";

export const GET_ZELLER_CUSTOMER = gql`
  query GetZellerCustomer($id: String!) {
    getZellerCustomer(id: $id) {
      id
      name
      email
      role
    }
  }
`;

export const LIST_ZELLER_CUSTOMERS = gql`
  query ListZellerCustomers(
    $filter: TableZellerCustomerFilterInput
    $limit: Int
    $nextToken: String
  ) {
    listZellerCustomers(filter: $filter, limit: $limit, nextToken: $nextToken) {
      items {
        id
        name
        email
        role
      }
      nextToken
    }
  }
`;
