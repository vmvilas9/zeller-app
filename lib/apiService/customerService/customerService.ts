import {ApolloQueryResult} from "@apollo/client";
import client from "../../network/client";
import {GET_ZELLER_CUSTOMER, LIST_ZELLER_CUSTOMERS} from "./queries";
import {TableZellerCustomerFilterInput, ZellerCustomer, ZellerCustomerConnection} from "./types";

export async function fetchZellerCustomer(id: string): Promise<ZellerCustomer | undefined> {
    try {
        const result: ApolloQueryResult<{ getZellerCustomer: ZellerCustomer }> = await client.query({
            query: GET_ZELLER_CUSTOMER,
            variables: {id},
        });
        return result.data?.getZellerCustomer;
    } catch (error) {
        console.error("Error fetching customer:", error);
        return undefined;
    }
}

export async function listZellerCustomers (
    filter?: TableZellerCustomerFilterInput,
    limit: number = 50,
    nextToken?: string
): Promise<ZellerCustomerConnection> {
    try {
        const result: ApolloQueryResult<{
            listZellerCustomers: ZellerCustomerConnection;
        }> = await client.query({
            query: LIST_ZELLER_CUSTOMERS,
            variables: {filter, limit, nextToken},
        });

        return result.data?.listZellerCustomers || {items: [], nextToken: null};
    } catch (error) {
        console.error("Error listing customers:", error);
        throw error;
    }
}
