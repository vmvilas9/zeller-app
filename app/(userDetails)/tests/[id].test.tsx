import React from "react";
import {render, waitFor} from "@testing-library/react-native";
import {useLocalSearchParams} from "expo-router";
import {fetchZellerCustomer} from "@/lib/apiService/customerService/customerService";
import CustomerDetail from "../[id]";

// Mock the necessary dependencies
jest.mock("expo-router", () => ({
    useLocalSearchParams: jest.fn(),
}));

jest.mock("@/lib/apiService/customerService/customerService", () => ({
    fetchZellerCustomer: jest.fn(),
}));

describe("CustomerDetail", () => {
    const mockCustomer = {
        id: "123",
        name: "John Doe",
        email: "johndoe@example.com",
        role: "Admin",
    };

    beforeEach(() => {
        // Clear all mocks before each test
        jest.clearAllMocks();
    });

    it('renders correctly', async () => {
        // Mock the params to simulate fetching for a specific customer ID
        (useLocalSearchParams as jest.Mock).mockReturnValue({id: "123"});

        // Mock fetchZellerCustomer to return a customer object
        (fetchZellerCustomer as jest.Mock).mockResolvedValueOnce(mockCustomer);
        const {toJSON, queryByTestId} = render(<CustomerDetail/>);
        await waitFor(() => expect(queryByTestId('activity-indicator')).toBeNull());
        expect(toJSON()).toMatchSnapshot();
    });

    it("renders loading indicator when data is being fetched", async () => {
        // Mock the params to simulate fetching for a specific customer ID
        (useLocalSearchParams as jest.Mock).mockReturnValue({id: "123"});

        // Mock fetchZellerCustomer to return a promise
        (fetchZellerCustomer as jest.Mock).mockResolvedValueOnce(undefined);

        const {findByTestId} = render(<CustomerDetail/>);

        const loader = await findByTestId('activity-indicator');

        // Check if the loading component is rendered
        expect(loader).toBeTruthy();
    });

    it("renders error message if customer data is not found", async () => {
        // Mock the params to simulate fetching for a specific customer ID
        (useLocalSearchParams as jest.Mock).mockReturnValue({id: "123"});

        // Mock fetchZellerCustomer to return undefined (simulating no customer found)
        (fetchZellerCustomer as jest.Mock).mockResolvedValueOnce(undefined);

        const {findByTestId, queryByTestId} = render(<CustomerDetail/>);

        // Wait for the loading state to be gone
        await waitFor(() => expect(queryByTestId('activity-indicator')).toBeNull());

        // Check if the "No customer found" message is rendered
        expect(findByTestId('no-data')).toBeTruthy();
    });

    it("renders customer details when data is fetched successfully", async () => {
        // Mock the params to simulate fetching for a specific customer ID
        (useLocalSearchParams as jest.Mock).mockReturnValue({id: "123"});

        // Mock fetchZellerCustomer to return a customer object
        (fetchZellerCustomer as jest.Mock).mockResolvedValueOnce(mockCustomer);

        const {findByTestId} = render(<CustomerDetail/>);

        const userDetails = await findByTestId('user-details');

        // Check if the customer name is rendered in UserDetails
        expect(userDetails).toBeTruthy();
    });

    it("handles errors in fetching customer data", async () => {
        // Mock the params to simulate fetching for a specific customer ID
        (useLocalSearchParams as jest.Mock).mockReturnValue({id: "123"});

        // Mock fetchZellerCustomer to throw an error
        (fetchZellerCustomer as jest.Mock).mockRejectedValueOnce(new Error("Error fetching customer"));

        const {findByTestId, queryByTestId} = render(<CustomerDetail/>);

        // Wait for the loading state to be gone
        await waitFor(() => expect(queryByTestId('activity-indicator')).toBeNull());

        // Check if the "No customer found" message is rendered
        expect(findByTestId('no-data')).toBeTruthy();
    });
});

