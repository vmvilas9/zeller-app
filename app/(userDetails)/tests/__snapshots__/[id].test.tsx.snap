// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerDetail renders correctly with default props 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
      },
      {
        "height": "100%",
        "padding": 20,
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
        },
        undefined,
      ]
    }
    testID="user-details"
  >
    <Text
      style={
        [
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      ID: 
      123
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      Name: 
      John Doe
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      Email: 
      johndoe@example.com
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      Role: 
      Admin
    </Text>
  </View>
</View>
`;
